<?xml version="1.0" encoding="UTF-8" ?>
<launch>
    <arg name="robot" default="X1"/>
    <arg name="robot_type" default="x1"/>
    <arg name="cmd_vel" default="cmd_vel"
         doc="Topic to which to publish velocity commands."/>
    <arg name="path_to_follow" default="path"
         doc="Path topic to follow."/>
    <arg name="points" default="points_slow_filtered"/>
    <arg name="keep_path" default="3."/>
    <arg name="map_frame" default="subt"/>
    <arg name="max_age" default="1.0"/>
    <arg name="obstacle_avoidance" default="false"/>
    <arg name="local_goal_dims" default="'xyz'"/>
    <arg name="goal_reached_dist" default="0.3"/>

    <node name="path_follower" pkg="naex" type="path_follower"
          respawn="true" respawn_delay="1.0" output="screen">

        <rosparam subst_value="true">
            map_frame: $(arg map_frame)
            odom_frame: $(arg robot)/odom
            robot_frame: $(arg robot)/base_footprint
            control_freq: 10.0
            goal_reached_dist: $(arg goal_reached_dist)
            goal_reached_angle: .inf
            use_path_theta: none
            max_age: $(arg max_age)
            obstacle_avoidance: $(arg obstacle_avoidance)
            <!--
            Max. path distances, tail is consumed first by reached goals.
            A higher value is used in the beginning to traverse across
            unobserved area around robot.
            -->
            max_path_dist: [2.0, 2.5, 2.5, 2.5]
            look_ahead: 0.8
            max_speed: 1.0
            allow_backward: true
            max_force_through_speed: 0.2
            max_angular_rate: 1.5
            keep_path: $(arg keep_path)
            keep_cloud_box: [[-3.0, 3.0], [-3.0, 3.0], [-2.0, 2.0]]
            clearance_box: [[-0.6, 0.6], [-0.45, 0.45], [0.2, 0.7]]
            show_clearance_pos: [-2, 2]
            min_points_obstacle: 3
            local_goal_dims: $(arg local_goal_dims)

            force_through_after: 10.
            backtrack_after: 20.
        </rosparam>

        <rosparam if="$(eval robot_type == 'dtr')" subst_value="true">
            clearance_box: [[-0.5, 0.5], [-0.4, 0.4], [0.2, 0.7]]
        </rosparam>
        <rosparam if="$(eval robot_type == 'jeanine')" subst_value="true">
            max_path_dist: [0.8, 1.5, 1.5, 1.5]
            max_speed: 1.0
            max_angular_rate: 1.5
            clearance_box: [[-0.4, 0.4], [-0.3, 0.3], [0.2, 0.6]]
        </rosparam>
        <rosparam if="$(eval robot_type == 'marv')" subst_value="true">
            allow_backward: true
            clearance_box: [[-0.55, 0.55], [-0.4, 0.4], [0.2, 0.7]]
        </rosparam>
        <rosparam if="$(eval robot_type == 'tradr')" subst_value="true">
            clearance_box: [[-0.5, 0.5], [-0.4, 0.4], [0.2, 0.6]]
        </rosparam>
        <rosparam if="$(eval 'x1' in robot_type)" subst_value="true">
            allow_backward: true
            clearance_box: [[-0.6, 0.6], [-0.45, 0.45], [0.2, 0.7]]
        </rosparam>
        <rosparam if="$(eval robot_type == 'x2')" subst_value="true">
            allow_backward: true
            clearance_box: [[-0.4, 0.4], [-0.3, 0.3], [0.15, 0.5]]
        </rosparam>

        <!-- Inputs -->
        <remap from="path" to="$(arg path_to_follow)"/>
        <remap from="cloud" to="$(arg points)"/>

        <!-- Outputs -->
        <remap from="clearance" to="clearance"/>
        <remap from="cmd_vel" to="$(arg cmd_vel)"/>
    </node>
</launch>
